pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,304 @@
+	kubeconfigContext := a.GetString(ParamKubeconfigContext)
+	if kubeconfigPath != "" {
+		configOverrides := &clientcmd.ConfigOverrides{}
+		if kubeconfigContext != "" {
+			configOverrides.CurrentContext = kubeconfigContext
+		}
+
+		restConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfigPath},
+			configOverrides).ClientConfig()
+	} else {
+		restConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	apiQPS := a.GetFloat64(ParamAPIQPS)
+	apiQPSBurstFactor := a.GetFloat64(ParamAPIQPSBurstFactor)
+	restConfig.RateLimiter = flowcontrol.NewTokenBucketRateLimiter(float32(apiQPS), int(apiQPS*apiQPSBurstFactor))
+	restConfig.UserAgent = UserAgent + "/" + options.Version
+
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return NewProvider(options, clientset)
+}
+
+// NewProvider returns a new k8s provider with the given k8s client configuration.
+func NewProvider(options gostatsd.CloudProviderOptions, clientset kubernetes.Interface) (gostatsd.CloudProvider, error) {
+	a := util.GetSubViper(options.Viper, "k8s")
+	a.SetDefault(ParamAPIQPS, DefaultAPIQPS)
+	a.SetDefault(ParamAPIQPSBurstFactor, DefaultAPIQPSBurstFactor)
+	a.SetDefault(ParamAnnotationTagWhitelist, DefaultAnnotationTagWhitelist)
+	a.SetDefault(ParamLabelTagWhitelist, DefaultLabelTagWhitelist)
+	a.SetDefault(ParamResyncPeriod, DefaultResyncPeriod)
+	a.SetDefault(ParamWatchCluster, DefaultWatchCluster)
+
+	// Create the tag whitelists
+	annotationWhitelist, err := stringListToRegex(a.GetStringSlice(ParamAnnotationTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+	labelWhitelist, err := stringListToRegex(a.GetStringSlice(ParamLabelTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up informers
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+
+	watchCluster := a.GetBool(ParamWatchCluster)
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+	if !watchCluster {
+		if options.NodeName == "" {
+			return nil, fmt.Errorf("'%s' set to false, and node name not supplied", ParamWatchCluster)
+		}
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fmt.Sprintf("spec.nodeName=%s", options.NodeName)
+		}
+	}
+
+	podsInf := core_v1inf.NewFilteredPodInformer(
+		clientset,
+		meta_v1.NamespaceAll,
+		a.GetDuration(ParamResyncPeriod),
+		indexers,
+		customWatchOptions)
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
+	// TODO: we should emit events to the k8s API to fit in with the k8s ecosystem
+
+	return &Provider{
+		logger:              options.Logger,
+		podsInf:             podsInf,
+		annotationWhitelist: annotationWhitelist,
+		labelWhitelist:      labelWhitelist,
+	}, nil
+}
+
+func (p *Provider) Run(ctx context.Context) {
+	p.podsInf.Run(ctx.Done())
+}
+
+func stringListToRegex(strList []string) (*regexp.Regexp, error) {
```

Comment:
When I suggested to make it a single regex I meant that the user should specify a single regex. We can take in a single string, simplifying our code and also making it more performant (1 match vs N), without limiting the user's ability to achieve what they want. I'm sorry I was not clear enough.

Reply:
Will capture groups actually work with this approach though? I remember you mentioned we could use capture groups to simplify how the tag is captured and processed, but I'm not sure how we would reference these in a multi-regex.

Reply:
When capturing is not needed/desired, a non-capturing group can be used. See https://github.com/google/re2/wiki/Syntax

Example: `(?:one|two|three)/(\w+)`

Reply:
We can also just mandate the use of a named capturing group. That might be a better option, actually. Who knows, maybe later we'll want to capture more things and then depending on order becomes problematic.

Reply:
I've gone with a sane default. If the regex matches and the "tag" capture group matches, use the tag. Otherwise use the regex match.